scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -354,41 +359,40 @@ class KRouter[A](
       }
     }
 
-    closestKnownNodesTask.flatMap { closestKnownNodes =>
-      if (closestKnownNodes.isEmpty) {
-        Task(logger.debug("Lookup finished without any nodes, as bootstrap nodes ")).as(Seq.empty)
-      } else {
-        val initalRequestState = closestKnownNodes.foldLeft(Map.empty[BitVector, RequestResult]) { (map, node) =>
-          map + (node.id -> RequestScheduled)
-        }
+    closestKnownNodesTask.map(NonEmptyList.fromList).flatMap {
+      case None =>
+        Task(logger.debug("Lookup finished without any nodes, as bootstrap nodes ")).as(Set.empty)
+
+      case Some(closestKnownNodes) =>
+        val initalRequestState: Map[BitVector, RequestResult] =
+          closestKnownNodes.toList.map(_.id -> RequestScheduled).toMap
         // All initial nodes are scheduled to request
-        val lookUpTask: Task[Seq[NodeRecord[A]]] = for {
+        val lookUpTask: Task[Set[NodeRecord[A]]] = for {
           // All initial nodes are scheduled to request
           state <- Ref.of[Task, Map[BitVector, RequestResult]](initalRequestState)
           // closestKnownNodes are constrained by alpha, it means there will be at most alpha independent recursive tasks
-          results <- Task.parTraverse(closestKnownNodes) { knownNode =>
-            recLookUp(List(knownNode), NonEmptyList.fromListUnsafe(closestKnownNodes.toList), state)
+          results <- Task.parTraverse(closestKnownNodes.toList) { knownNode =>
```

Comment:
I am thinking, we are constantly doing this dance `.toList`, `.toSet` and`(nodesFound ++ receivedNodes).distinct(xorOrder).sorted(xorOrder)` (in recLookupFunction)
Do you think it would make sense to use `NonEmptySet` ? (from what I rember it uses scala `SortedSet`, se we would have nicly ordered and diststinct results)

Reply:
That sounds like a great idea!

Reply:
Done.